core/src/main/java/io/grpc/internal/TransportTracer.java

```diff
@@ -0,0 +1,212 @@
+   * only one thread, but the resulting stats may be read from any thread.
+   */
+  public void reportStreamStarted() {
+    // @SuppressWarnings is not allowed on an increment statement, using another var here.
+    long newStreamsStarted = streamsStarted + 1;
+    streamsStarted = newStreamsStarted;
+    lastStreamCreatedTimeMsec = System.currentTimeMillis();
+  }
+
+  /**
+   * Reports that a keep alive message was sent.
+   */
+  public void reportKeepAliveSent() {
+    keepAlivesSent.incrementAndGet();
+  }
+
+  /**
+   * Registers a {@code Callable<Integer>} that can be used to poll for the remote flow control
+   * window size.
+   */
+  public void setRemoteFlowControlWindowPollable(Callable<Integer> remoteFlowControllPollable) {
+    this.remoteFlowControlPollable = remoteFlowControllPollable;
+  }
+
+  /**
+   * Registers a {@code Callable<Integer>} that can be used to poll for the local flow control
+   * window size.
+   */
+  public void setLocalFlowControlWindowPollable(Callable<Integer> localFlowControllPollable) {
+    this.localFlowControlPollable = localFlowControllPollable;
+  }
+
+  /**
+   * Returns the number of streams started on the transport.
+   */
+  public long getStreamsStarted() {
+    return streamsStarted;
+  }
+
+  /**
+   * Returns the number of streams ended successfully with an OK status.
+   */
+  public long getStreamsSucceeded() {
+    return streamsSucceeded.get();
+  }
+
+  /**
+   * Returns the number of streams completed with a non-OK status.
+   */
+  public long getStreamsFailed() {
+    return streamsFailed.get();
+  }
+
+  /**
+   * Returns the number of messages sent on the transport.
+   */
+  public long getMessagesSent() {
+    return messagesSent.get();
+  }
+
+  /**
+   * Returns the number of messages received on the transport.
+   */
+  public long getMessagesReceived() {
+    return messagesReceived.get();
+  }
+
+  /**
+   * Returns the number of keep alive messages sent on the transport.
+   */
+  public long getKeepAlivesSent() {
+    return keepAlivesSent.get();
+  }
+
+  /**
+   * Returns the last time a stream was created as millis since Unix epoch.
+   */
+  public long getLastStreamCreatedTimeMsec() {
+    return lastStreamCreatedTimeMsec;
+  }
+
+  /**
+   * Returns the last time a message was sent as millis since Unix epoch.
+   */
+  public long getLastMessageSentTimeMsec() {
+    return lastMessageSentTimeMsec.get();
+  }
+
+  /**
+   * Returns the last time a message was received as millis since Unix epoch.
+   */
+  public long getLastMessageReceivedTimeMsec() {
+    return lastMessageReceivedTimeMsec.get();
+  }
+
+  /**
+   * Returns the remote flow control window as reported by the callback of
+   * {@link #setRemoteFlowControlWindowPollable}. Returns -1 if no callback was registered.
+   */
+  public int getRemoteFlowControlWindow() {
+    if (remoteFlowControlPollable == null) {
```

Comment:
store the poller in the local?   that will avoid the double volatile read.

Reply:
Done